src/constraints/battery_degradation.jl

```diff
@@ -0,0 +1,207 @@
+        # 80% of the installed cost since we are not replacing the entire battery
+        f = strategy == "augmentation" ? 0.8 : 1.0
+        p.s.storage.attr[b].degradation.maintenance_cost_per_kwh = [ f * 
+            p.s.storage.attr[b].installed_cost_per_kwh * pwf(d) for d in days[1:end-1]
+        ]
+    end
+
+    @assert(length(p.s.storage.attr[b].degradation.maintenance_cost_per_kwh) == length(days) - 1,
+        "The degradation maintenance_cost_per_kwh must have a length of $(length(days)-1)."
+    )
+
+    @variable(m, SOH[days])
+
+    add_degradation_variables(m, p)
+    constrain_degradation_variables(m, p, b=b)
+
+    @constraint(m, [d in 2:days[end]],
+        SOH[d] == SOH[d-1] - p.hours_per_timestep * (
+            p.s.storage.attr[b].degradation.calendar_fade_coefficient * time_exponent * 
+            m[:Eavg][d-1] * d^(time_exponent-1) + 
+            p.s.storage.attr[b].degradation.cycle_fade_coefficient * m[:EFC][d-1]
+        )
+    )
+    # NOTE SOH can be negative
+
+    @constraint(m, SOH[1] == m[:dvStorageEnergy][b])
+    # NOTE SOH is _not_ normalized, and has units of kWh
+
+    if strategy == "replacement"
+        #=
+        When the battery degrades to 80% of purchased capacity it is replaced.
+        Multiple replacements could be necessary within the analysis period.
+        (The SOH is calculated for the analysis period, but not for multiple batteries.)
+        So we construct a cost as a function of months that accounts for the number of replacements.
+        (We use months instead of days to reduce the number of integer variables required).
+
+        The replacment cost in a given month is:
+        1. the maintenance_cost_per_kwh in (approximately) the 15th day of the month multiplied with
+        2. the number of replacements required given the first month that the battery must be replaced.
+        The number of months is analysis_years * 12.
+        The first month that the battery is replaced is determined by d_0p8, which is the integer 
+        number of days that the SOH is at least 80% of the purchased capacity.
+        We define a binary for each month and only allow one month to be chosen.
+        =#
+        
+        # define d_0p8
+        @warn "Adding binary and indicator constraints for 
+         ElectricStorage.degradation.maintenance_strategy = \"replacement\". 
+         Not all solvers support indicators and some are slow with integers."
+        @variable(m, soh_indicator[days], Bin)
+        @constraint(m, [d in days],
+            soh_indicator[d] => {SOH[d] >= 0.8*m[:dvStorageEnergy][b]}
+        )
+        @expression(m, d_0p8, sum(soh_indicator[d] for d in days))
+
+        # define binaries for the finding the month that battery must be replaced
+        months = 1:p.s.financial.analysis_years*12
+        @variable(m, bmth[months], Bin)
+        # can only pick one month
+        @constraint(m, sum(bmth[mth] for mth in months) == 1)
```

Comment:
I think that if we make it <= 1 then it will always choose bmth[mth] = 0 forall mth in months (because that results in zero replacement costs). However, I see your point that this effectively forces choosing at least the last month (because it has the lowest cost). So maybe the fix is to make the last value in the cost curve * number of replacements (`c` below) equal to zero? That way if the SOH is greater or equal to 80% of initial capacity for all days then there is zero replacement cost.

Reply:
I see now, we don't have anything else forcing replacement of the battery.  Leaving the sense as "==" and changing the right-hand side to soh_indicator[length(days)] will do that for us, and that way if SOH stays above 80% of capacity, we just don't replace.

Reply:
will it make the problem harder to solve by having sum of binaries == binary? do we get the same affect by making the last cost in `c` zero?

Reply:
I ran the test case both ways.  The test case showed no noticeable difference in solution time (it's a single constraint, and not too dense related to the number of decision variables in REopt; it's sum of length(months) vs length(months)+1 binaries = constant, so not a big difference), but if the objective value for replacement in the last month is set to zero the end SOH falls a little below 80%.  

Ultimately I think this is a cleaner approach but I'm open to either running more cases to test it if you're concerned.

Reply:
ok thanks for testing! yes let's go with your suggested approach